pkg/subctl/cmd/cloud/gcp/gcp.go

```diff
@@ -0,0 +1,179 @@
+			"and region from (takes precedence over the specific flags)")
+
+	dirname, err := os.UserHomeDir()
+	if err != nil {
+		utils.ExitOnError("failed to find home directory", err)
+	}
+
+	defaultCredentials := filepath.FromSlash(fmt.Sprintf("%s/.gcp/osServiceAccount.json", dirname))
+	command.Flags().StringVar(&credentialsFile, "credentials", defaultCredentials, "GCP credentials configuration file")
+}
+
+// RunOnGCP runs the given function on GCP, supplying it with a cloud instance connected to GCP and a reporter that writes to CLI.
+// The functions makes sure that infraID and region are specified, and extracts the credentials from a secret in order to connect to GCP.
+func RunOnGCP(gwInstanceType, kubeConfig, kubeContext string, dedicatedGWNodes bool,
+	function func(cloud api.Cloud, gwDeployer api.GatewayDeployer, reporter api.Reporter) error) error {
+	if ocpMetadataFile != "" {
+		err := initializeFlagsFromOCPMetadata(ocpMetadataFile)
+		utils.ExitOnError("Failed to read GCP Cluster information from OCP metadata file", err)
+	} else {
+		utils.ExpectFlag(infraIDFlag, infraID)
+		utils.ExpectFlag(regionFlag, region)
+		utils.ExpectFlag(projectIDFlag, region)
+	}
+
+	reporter := cloudutils.NewCLIReporter()
+	reporter.Started("Retrieving GCP credentials from your GCP configuration")
+	creds, err := getGCPCredentials()
+	utils.ExitOnError("Failed to get GCP credentials", err)
+	reporter.Succeeded("")
+
+	reporter.Started("Initializing GCP connectivity")
+
+	options := []option.ClientOption{
+		option.WithCredentials(creds),
+		option.WithUserAgent("open-cluster-management.io submarineraddon/v1"),
+	}
+
+	gcpClient, err := gcpClientIface.NewClient(projectID, options)
+	utils.ExitOnError("Failed to initialize a GCP Client", err)
+
+	k8sConfig, err := restconfig.ForCluster(kubeConfig, kubeContext)
+	utils.ExitOnError("Failed to initialize a Kubernetes config", err)
+
+	clientSet, err := kubernetes.NewForConfig(k8sConfig)
+	utils.ExitOnError("Failed to create Kubernetes client", err)
+
+	k8sClientSet, err := k8s.NewK8sInterface(clientSet)
+	utils.ExitOnError("Failed to create cloud-prepare k8s client", err)
+
+	restMapper, err := util.BuildRestMapper(k8sConfig)
+	utils.ExitOnError("Failed to create restmapper", err)
+
+	dynamicClient, err := dynamic.NewForConfig(k8sConfig)
+	utils.ExitOnError("Failed to create dynamic client", err)
+
+	gcpCloud := cloudpreparegcp.NewCloud(projectID, infraID, region, gcpClient)
+	msDeployer := ocp.NewK8sMachinesetDeployer(restMapper, dynamicClient)
+	// TODO: Ideally we should be able to specify the image for GWNode, but it was seen that
+	// with certain images, the instance is not coming up. Needs to be investigated further.
+	gwDeployer, err := cloudpreparegcp.NewOcpGatewayDeployer(gcpCloud, msDeployer, gwInstanceType, "", dedicatedGWNodes, k8sClientSet)
+	utils.ExitOnError("Failed to initialize a GatewayDeployer config", err)
+
+	return function(gcpCloud, gwDeployer, reporter)
```

Comment:
The function still returns an error here so it's still inconsistent but I guess it doesn't matter.